pretyping/cases.ml

```diff
@@ -1055,26 +1055,20 @@ let add_assert_false_case pb tomatch =
 let adjust_impossible_cases sigma pb pred tomatch submat =
   match submat with
   | [] ->
-    (* FIXME: This breaks if using evar-insensitive primitives. In particular,
-       this means that the Evd.define below may redefine an already defined
-       evar. See e.g. first definition of test for bug #3388. *)
-    let pred = EConstr.Unsafe.to_constr pred in
-    begin match Constr.kind pred with
+    begin match EConstr.kind sigma pred with
     | Evar (evk, _) ->
-      let EvarInfo evi = Evd.find sigma evk in
+      let evi = Evd.find_undefined sigma evk in
       if snd (Evd.evar_source evi) == Evar_kinds.ImpossibleCase then
-        let sigma =
-          if not (Evd.is_defined sigma evk) then
-            let sigma, default = coq_unit_judge !!(pb.env) sigma in
-            let sigma = Evd.define evk default.uj_type sigma in
-            sigma
-          else sigma
-        in
+        let sigma, default = coq_unit_judge !!(pb.env) sigma in
+        let sigma = Evd.define evk default.uj_type sigma in
         sigma, add_assert_false_case pb tomatch
       else
         sigma, submat
     | _ ->
-        sigma, submat
+      let sigma', default = coq_unit_judge !!(pb.env) sigma in
+      if EConstr.eq_constr_nounivs sigma' pred default.uj_type
```

Comment:
This is really ugly, we should track the evars instantiated to absurd case instead. But well for the time being I guess this can stay.